**MASTER'S DEGREE IN APPLIED ARTIFICIAL INTELLIGENCE**.

**Course: TC5035.10 - Integration project**

Tecnológico de Monterrey

Professor Dr. Horacio Martínez Alfaro

**Week 4, Activity 3**

Baseline

---

*   **Student:** Jorge Antonio Acosta Rosales
*   **Registration:** A01381552
*   **Student:** Israel Luján González
*   **Registration:** A01794693

In [14]:
import os
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix
import joblib


In [15]:
# Load variables from .env file for local development or get it from the environment variables in production
try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    pass  # Safe fallback if it's not available in prod

base_path = os.getenv("LSM_BASE")
if not base_path:
    raise ValueError("❌ Environment variable 'LSM_BASE' is not set!")

gestures_csv = os.path.join(base_path, "data", "processed", "metadata", "gestures_balanced.csv")

In [16]:
df = pd.read_csv(gestures_csv)
df.head()

PC1       PC2       PC3       PC4       PC5       PC6       PC7  \
0  0.567907  0.114514  0.210044 -0.047475  0.088927 -0.411519  0.339483   
1  0.297408  0.792936 -0.259329  0.508329 -0.105111  0.353935  0.091841   
2 -0.257034  0.550256 -0.625594 -0.684124  0.220932 -0.449982  0.089296   
3  0.211764  0.618572 -0.297504  0.296913 -0.474989  0.458511 -0.047803   
4 -0.376056  0.095322 -0.195378  0.588176 -0.142988  0.070855 -0.682290   

        PC8       PC9      PC10      PC11      PC12      PC13      PC14  \
0 -0.719156  0.404330  0.374506  0.035606 -0.028653  0.091038  0.015528   
1 -0.347120 -0.088993  0.356542  0.118040 -0.007878 -0.096861  0.233572   
2  0.215863 -0.307653  0.219697  0.025681  0.016383 -0.226053  0.183939   
3 -0.355956 -0.034894  0.365687  0.057215 -0.258740 -0.172323  0.219386   
4  0.373771 -0.158868  0.325924  0.028607 -0.170780 -0.133886 -0.195853   

       PC15 label  
0 -0.097783     A  
1 -0.018293     A  
2  0.005376     A  
3 -0.043902     A  
4  0.003899     A

Split the dataset into training and test sets


In [17]:
X = df.drop(columns=["label"])
y = df["label"]

# Split dataset for training and testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

In [18]:
clf = RandomForestClassifier(n_estimators=100, random_state=42)
clf.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [22]:
y_pred = clf.predict(X_test)
print("Confusion Matrix:")
cm = confusion_matrix(y_test, y_pred)
print(cm)

Confusion Matrix:
[[25  0  1  0  0  0  0  0  0  0  0  1  0  0  0  1  0  1  0  0  1  0  0  0
   0  0  0  1]
 [ 0 29  0  0  1  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
   0  0  0  0]
 [ 0  0 28  2  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
   0  0  0  0]
 [ 0  1  0 30  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
   0  0  0  0]
 [ 0  0  0  1 29  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
   0  0  0  0]
 [ 0  4  0  0  0 27  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
   0  0  0  0]
 [ 0  0  0  0  0  0 31  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
   0  0  0  0]
 [ 0  0  0  0  0  0  0 31  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
   0  0  0  0]
 [ 0  0  0  0  0  0  0  0 31  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
   0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0 31  0  0  0  0  0  0  0  0  0  0  0  0  0  0
   0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0 26  0  1  1  1  0  0  0  0  0  0  0  1  1
   0  0  0  0]
 [ 

In [24]:
labels = clf.classes_

dm_cm = pd.DataFrame(cm, index=labels, columns=labels)
dm_cm.head(28)

A   B   C   D   E   F   G   H   I   J  ...   R  RR   S   T   U   V   W  \
A   25   0   1   0   0   0   0   0   0   0  ...   0   0   1   0   0   0   0   
B    0  29   0   0   1   1   0   0   0   0  ...   0   0   0   0   0   0   0   
C    0   0  28   2   1   0   0   0   0   0  ...   0   0   0   0   0   0   0   
D    0   1   0  30   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   
E    0   0   0   1  29   0   0   1   0   0  ...   0   0   0   0   0   0   0   
F    0   4   0   0   0  27   0   0   0   0  ...   0   0   0   0   0   0   0   
G    0   0   0   0   0   0  31   0   0   0  ...   0   0   0   0   0   0   0   
H    0   0   0   0   0   0   0  31   0   0  ...   0   0   0   0   0   0   0   
I    0   0   0   0   0   0   0   0  31   0  ...   0   0   0   0   0   0   0   
J    0   0   0   0   0   0   0   0   0  31  ...   0   0   0   0   0   0   0   
K    0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   1   1   0   
L    0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   
LL   0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   
M    0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   
N    0   0   0   0   0   0   0   0   1   0  ...   0   0   0   1   0   0   0   
O    0   0   0   3   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   
P    0   0   0   1   0   0   0   0   0   0  ...   0   0   2   0   0   0   0   
Q    0   0   0   0   0   0   0   0   0   0  ...   0   0   2   0   0   0   0   
R    0   0   0   0   0   0   0   0   0   0  ...  30   0   0   0   0   0   0   
RR   0   0   0   0   0   0   0   0   0   0  ...   0  31   0   0   0   0   0   
S    2   0   0   1   1   0   1   0   0   0  ...   0   0  23   1   0   0   0   
T    0   0   0   0   0   0   0   0   1   0  ...   0   0   0  29   0   0   0   
U    0   0   0   1   0   0   0   0   0   0  ...   0   0   0   0  28   2   0   
V    0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   1  30   0   
W    0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0  31   
X    0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   
Y    0   0   0   0   0   0   0   0   1   0  ...   0   0   0   0   0   0   0   
Z    0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   

     X   Y   Z  
A    0   0   1  
B    0   0   0  
C    0   0   0  
D    0   0   0  
E    0   0   0  
F    0   0   0  
G    0   0   0  
H    0   0   0  
I    0   0   0  
J    0   0   0  
K    0   0   0  
L    0   0   0  
LL   0   0   0  
M    0   0   0  
N    1   0   0  
O    0   0   0  
P    0   0   1  
Q    0   0   0  
R    0   0   0  
RR   0   0   0  
S    0   1   0  
T    0   0   1  
U    0   0   0  
V    0   0   0  
W    0   0   0  
X   31   0   0  
Y    0  30   0  
Z    0   0  30  

[28 rows x 28 columns]

In [21]:
print("Classification Report:")
print(classification_report(y_test, y_pred))

Classification Report:
              precision    recall  f1-score   support

           A       0.93      0.81      0.86        31
           B       0.85      0.94      0.89        31
           C       0.97      0.90      0.93        31
           D       0.77      0.97      0.86        31
           E       0.91      0.94      0.92        31
           F       0.96      0.87      0.92        31
           G       0.97      1.00      0.98        31
           H       0.97      1.00      0.98        31
           I       0.91      1.00      0.95        31
           J       1.00      1.00      1.00        31
           K       0.87      0.84      0.85        31
           L       0.97      0.94      0.95        31
          LL       0.89      1.00      0.94        31
           M       0.94      1.00      0.97        31
           N       0.93      0.81      0.86        31
           O       0.90      0.90      0.90        31
           P       1.00      0.77      0.87        31
    

Here are the key conclusions drawn from the classification report:

1. **Strong Overall Performance**

   * The overall accuracy is 93% (868 correctly classified samples out of 868).
   * The macro and weighted averages for precision, recall, and F1-score are all approximately 0.93, indicating balanced performance across all classes.

2. **Perfectly Classified Classes**

   * **J, RR, and W** achieved perfect scores (precision, recall, and F1 = 1.00), meaning they were always correctly identified with no confusion. This suggests these letters are highly distinctive or well-represented in the training data.

3. **High-Performing Classes (F1 ≥ 0.95)**

   * G, H, I, L, M, Q, R, X, Y, and Z show excellent performance, demonstrating the model reliably recognizes these letters.

4. **Classes with Precision vs. Recall Imbalance**

   * **D**: very high recall (0.97) but low precision (0.77), meaning most real D samples are correctly identified, but other letters are often misclassified as D (high false positives).
   * **F** and **K**: have recall scores below 90% or precision around 85–87%, indicating some confusion with neighboring classes.

5. **Classes Needing Improvement**

   * **P**: recall is only 0.77, meaning 7 out of 31 real P samples are misclassified.
   * **S**: precision of 0.82 and recall of 0.74 make it the most problematic class, with both false positives and false negatives.
   * **N** and **A** also show slightly lower recall values (0.81).

6. **Recommendations for Optimization**

   * Examine the confusion matrix to identify which specific classes are being confused with P, S, and D.
   * Increase or balance the dataset (e.g., via data augmentation) for low-recall classes.
   * Consider adjusting class weights (`class_weight`) or prediction thresholds to reduce critical errors.
   * Engineer additional features (e.g., specific angles or distances between landmarks) to better distinguish confusing letters.

**In summary**, the model performs very well for most letters (many with F1-scores ≥ 0.95), but extra attention should be given to P, S, and D to reduce false positives and false negatives, either by improving data balance, tuning hyperparameters, or refining feature extraction.
